OpenRA.Platforms.Default/DummySoundEngine.cs

```diff
@@ -0,0 +1,89 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.IO;
+
+namespace OpenRA.Platforms.Default
+{
+	sealed class DummySoundEngine : ISoundEngine
+	{
+		public SoundDevice[] AvailableDevices()
+		{
+			var defaultDevices = new[]
+			{
+				new SoundDevice(null, "Default Output"),
+			};
+
+			return defaultDevices;
+		}
+
+		public DummySoundEngine(string deviceName)
+		{
+		}
+
+		public ISoundSource AddSoundSourceFromMemory(byte[] data, int channels, int sampleBits, int sampleRate)
+		{
+			return null;
+		}
+
+		public ISound Play2D(ISoundSource soundSource, bool loop, bool relative, WPos pos, float volume, bool attenuateVolume)
+		{
+			return null;
```

Comment:
Are we sure we can return null here? Looks like it will introduce NullReferenceException where `ISound` is used.

Reply:
I'm not certain about this; however during my testing (which included starting a game and doing a few actions), it seemed to work as intended. I might quickly find which actions call that function and try them - there is a chance none of the actions I've tested called it.

Reply:
I tried again seems to work fine. One thing I haven't tried is the videos since I don't own a copy of the original game. This line is the only call which expects a return value:
https://github.com/OpenRA/OpenRA/blob/d5588c51ede7c29c42ab6158f9c56186ccbc75dd/OpenRA.Game/Sound/Sound.cs#L165

Reply:
Take a look where `ISound` is used in the code. There are several places where it takes for granted that its not null. For example here.
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Sound/AmbientSound.cs#L90

Reply:
This can be solved by adding a dummy sound class. We used to have this in the distant past:

https://github.com/OpenRA/OpenRA/blob/eed38b3da4f45d7fa0576d183646496723c136fc/OpenRA.Game/Sound.cs#L721-L756

Reply:
I logged `s` just before it is added to the array and it is indeed `null`. However this doesn't have any affect on the game. I understand that it isn't good practice, so what do you suggest I return instead?

Reply:
Actually I noticed what may be a regression: When trying to force-attack one of your own units, the unit will be selected rather than attacked.

Reply:
I'm having quite a bit of trouble extracting the movies to test them. Any idea how I'd go about doing it (Linux)?

Reply:
Mount the iso at `/media/openra` and let the game do it for you through the "Manage Content" menu.

Reply:
Thanks, I'll try that.

Reply:
There seems to be two ISOs (`Allies` and `Soviets`). Should I mount both?

Reply:
I tried the videos and they were static (obviously a regression), so I've replaced mine with your implementation. Should I keep the name `DummySoundEngine` or use `NullSoundEngine`? The latter sounds better to me.

Reply:
BTW I was wondering - what was the reason for the previous implementation being removed?

Reply:
Hmm, this seems to be quite a bit more complex than first meets the eye. The videos seem to rely on the sound source to play, so the `Play2D` function is going to have to be taken from the OpenAL sound engine and modified accordingly.

Reply:
I've found the offending line:
https://github.com/OpenRA/OpenRA/blob/ba282865f174af363dba609374abc48d08cfe17b/OpenRA.Mods.Common/Widgets/VqaPlayerWidget.cs#L107
Hard-coding `video.HasAudio` to `false` fixed the problem. Would extending the `if` statement to check if the sound system is a dummy/null or not be an acceptable solution?

Reply:
@pchote Have you had a chance to review the above question?

Reply:
I have not, and probably won't for quite some time sorry. Real life commitments mean that I do not have time to spare for PR reviews right now.

Reply:
I understand - all the best.

Reply:
@teinarss Is it possible for you to take over for pchote in this regard?

Reply:
I've fixed the issue by checking if the sound engine is a dummy or not. Couldn't think of a very creative way of checking though.